### PR TITLE
Add interhand 3D heads and test codes.

### PR DESCRIPTION
**Pull request by zengwang430521**

### CODE REVIEW

File: mmpose/models/losses/classfication_loss.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..registry import LOSSES
+
+
+@LOSSES.register_module()
+class BCELoss(nn.Module):
+    """Binary Cross Entropy loss ."""
```
Comment (jin-s13) [inline mmpose/models/losses/classfication_loss.py:9]: del the space before `.`

Comment (zengwang430521) [reply mmpose/models/losses/classfication_loss.py:9]: OK, I will fix it

File: mmpose/models/keypoint_heads/interhand_3d_head.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            flip_pairs (None | list[tuple()):
+                Pairs of keypoints which are mirrored.
+        """
+
+        output = self.forward(x)
+
+        if flip_pairs is not None:
+            N, K, D, H, W = output.shape
+            # reshape 3D heatmap to 2D heatmap
+            output = output.reshape(N, K * D, H, W)
```
Comment (liuxin9608) [inline mmpose/models/keypoint_heads/interhand_3d_head.py:117]: N, K, D*H, W ?

File: mmpose/models/keypoint_heads/interhand_3d_head.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+            # reshape back to 3D heatmap
+            output_heatmap = output_heatmap.reshape(N, K, D, H, W)
+            # feature is not aligned, shift flipped heatmap for higher accuracy
+            if self.test_cfg.get('shift_heatmap', False):
+                # output_heatmap[:, :, :, 1:] = output_heatmap[:, :, :, :-1]
+                output_heatmap[..., 1:] = output_heatmap[..., :-1]
+        else:
+            output_heatmap = output.detach().cpu().numpy()
+        return output_heatmap
+
```
Comment (liuxin9608) [inline mmpose/models/keypoint_heads/interhand_3d_head.py:132]: decode keypoints

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+    preds = np.where(np.tile(maxvals, (1, 1, 3)) > 0.0, preds, -1)
+    return preds, maxvals
+
+
+@HEADS.register_module()
+class HeatMap3DHead(TopDownSimpleHead):
+    """3D heatmap head of paper ref: Gyeongsik Moon. "InterHand2.6M: A Dataset
+    and Baseline for 3D Interacting Hand Pose Estimation from a Single RGB
+    Image" HeatMap3DHead is a variant of TopDownSimpleHead, and is consisted of
```
Comment (jin-s13) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:50]: consisted -> composed

Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py]: fixed

File: mmpose/models/keypoint_heads/multilabel_classification_head.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    def inference_model(self, x, flip_pairs=None):
+        """Inference function.
+
+        Returns:
+            output_labels (np.ndarray): Output labels.
+
+        Args:
+            x (torch.Tensor[NxC]): Input features vector.
+            flip_pairs (None | list[tuple()):
```
Comment (jin-s13) [inline mmpose/models/keypoint_heads/multilabel_classification_head.py:105]: list[]

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+            flip_pairs (None | list[tuple()):
+                Pairs of keypoints which are mirrored.
+        """
+
+        output = self.forward(x)
+
+        if flip_pairs is not None:
+            N, K, D, H, W = output.shape
+            # reshape 3D heatmap to 2D heatmap
+            output = output.reshape(N, K, D * H, W)
```
Comment (jin-s13) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:208]: double check line 131

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+from ..registry import HEADS
+from .top_down_simple_head import TopDownSimpleHead
+
+
+def _get_max_preds_3d(heatmaps):
+    """Get keypoint predictions from 3D score maps.
+
+    Note:
+        batch_size: N
+        num_keypoints: K
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:14]: num keypoints

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        ('heatmaps should be numpy.ndarray')
+    assert heatmaps.ndim == 5, 'heatmaps should be 5-ndim'
+
+    N, K, D, H, W = heatmaps.shape
+    heatmaps_reshaped = heatmaps.reshape((N, K, -1))
+    idx = np.argmax(heatmaps_reshaped, 2).reshape((N, K, 1))
+    maxvals = np.amax(heatmaps_reshaped, 2).reshape((N, K, 1))
+
+    preds = np.tile(idx, (1, 1, 3)).astype(np.float32)
+    _idx = idx.copy()[:, :, 0]
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:37]: copy is not needed here.
```python
_idx = idx[..., 0]
```

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    assert isinstance(heatmaps, np.ndarray), \
+        ('heatmaps should be numpy.ndarray')
+    assert heatmaps.ndim == 5, 'heatmaps should be 5-ndim'
+
+    N, K, D, H, W = heatmaps.shape
+    heatmaps_reshaped = heatmaps.reshape((N, K, -1))
+    idx = np.argmax(heatmaps_reshaped, 2).reshape((N, K, 1))
+    maxvals = np.amax(heatmaps_reshaped, 2).reshape((N, K, 1))
+
+    preds = np.tile(idx, (1, 1, 3)).astype(np.float32)
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:36]: In L38-40, values in preds all are overriden. So,
```python
preds = np.zeros((N, K, 3), dtype=np.float32)
```

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+        if 'bbox_id' in img_metas[0]:
+            bbox_ids = []
+        else:
+            bbox_ids = None
+
+        center = np.zeros((batch_size, 2), dtype=np.float32)
+        scale = np.zeros((batch_size, 2), dtype=np.float32)
+        image_paths = []
+        score = np.ones(batch_size)
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:158]: dtype

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            preds[i, :, :2] = transform_preds(preds[i, :, :2], center[i],
+                                              scale[i], [W, H])
+
+        all_preds = np.zeros((batch_size, preds.shape[1], 3), dtype=np.float32)
+        all_boxes = np.zeros((batch_size, 6), dtype=np.float32)
+        all_preds[:, :, 0:2] = preds[:, :, 0:2]
+        all_preds[:, :, 2:3] = maxvals
+        all_boxes[:, 0:2] = center[:, 0:2]
+        all_boxes[:, 2:4] = scale[:, 0:2]
+        all_boxes[:, 4] = np.prod(scale * 200.0, axis=1)
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:181]: Add a note for the magic number 200

File: mmpose/models/keypoint_heads/multilabel_classification_head.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+        return labels
+
+    def decode(self, img_metas, output, **kwargs):
+        """Decode keypoints from heatmaps.
+
+        Args:
+            img_metas (list(dict)): Information about data augmentation
+                By default this includes:
+                - "image_file: path to the image file
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/multilabel_classification_head.py:125]: unpaired ` " `

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+    def forward(self, x):
+        """Forward function."""
+        x = super().forward(x)
+        N, C, H, W = x.shape
+        # reshape the 2D heatmap to 3D heatmap
+        x = x.reshape(N, C // self.depth_size, self.depth_size, H, W)
+        return x
+
+    def decode(self, img_metas, output, **kwargs):
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_3d_head.py:134]: Will it error if `kwargs` is removed?

File: mmpose/models/keypoint_heads/multilabel_classification_head.py (hunk 5 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/multilabel_classification_head.py:105]: fixed.

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+            # 2D heatmap flip
+            output_heatmap = flip_back(
+                output.detach().cpu().numpy(),
+                flip_pairs,
+                target_type=self.target_type)
+            # reshape back to 3D heatmap
+            output_heatmap = output_heatmap.reshape(N, K, D, H, W)
+            # feature is not aligned, shift flipped heatmap for higher accuracy
+            if self.test_cfg.get('shift_heatmap', False):
+                # output_heatmap[:, :, :, 1:] = output_heatmap[:, :, :, :-1]
```
[the remark this answers is not in the document]
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:218]: fixed

File: mmpose/models/keypoint_heads/heatmap_3d_head.py (hunk 6 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:208]: Line 131 is right.
Both reshaping 3D heatmap to (N, K, D * H, W) or (N, K * D, H, W) is right here. I do not know which is better here.

File: mmpose/models/keypoint_heads/heatmap_3d_head.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+from mmpose.core.post_processing import flip_back, transform_preds
+from ..registry import HEADS
+from .top_down_simple_head import TopDownSimpleHead
+
+
+def _get_max_preds_3d(heatmaps):
+    """Get keypoint predictions from 3D score maps.
+
+    Note:
+        batch_size: N
```
[the remark this answers is not in the document]
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:13]: fixed.

File: mmpose/models/keypoint_heads/heatmap_3d_head.py (hunk 7 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:14]: fixed

File: mmpose/models/keypoint_heads/heatmap_3d_head.py (hunk 8 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:37]: fixed

File: mmpose/models/keypoint_heads/heatmap_3d_head.py (hunk 9 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:36]: OK, fixed

File: mmpose/models/keypoint_heads/heatmap_3d_head.py (hunk 10 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:158]: fixed.

File: mmpose/models/keypoint_heads/multilabel_classification_head.py (hunk 12 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/multilabel_classification_head.py:125]: OK, fixed

File: mmpose/models/keypoint_heads/multilabel_classification_head.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    def get_accuracy(self, output, target, target_weight):
+        """Calculate accuracy for classification.
+
+        Note:
+            batch_size: N
+
+        Args:
+            output (torch.Tensor[N, 2]): Output hand visibility.
+            target (torch.Tensor[N, 2]): Target hand visibility.
+            target_weight (torch.Tensor[N, K, 2]):
```
[the remark this answers is not in the document]
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/multilabel_classification_head.py:87]: I will modify the codes to utilize target weight.

File: mmpose/models/keypoint_heads/heatmap_3d_head.py (hunk 13 above)
Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:134]: Some heads need extra information for decode function, for example img_size. So I keep _kwargs_ to make the decode function in the same format.

Comment (innerlee) [reply mmpose/models/keypoint_heads/heatmap_3d_head.py:134]: Okay. Lets keep it for now.

File: mmpose/models/keypoint_heads/heatmap_1d_head.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        """Forward function."""
+        heatmap1d = self.fc(x)
+        value = self.soft_argmax_1d(heatmap1d).view(-1, 1)
+        return value
+
+    def get_loss(self, output, target, target_weight):
+        """Calculate regression loss of heatmap.
+
+        Note:
+            batch_size: N
```
Comment (innerlee) [inline mmpose/models/keypoint_heads/heatmap_1d_head.py:71]: batch size

Comment (zengwang430521) [reply mmpose/models/keypoint_heads/heatmap_1d_head.py:71]: fixed.

**Review by innerlee — APPROVED** [review]

**Pull request merged**